graph/preprocessor.go

```diff
@@ -0,0 +1,211 @@
+func addAliasFromRemote(preTask *PreTask, url string) error {
+	remoteClient := http.Client{
+		Timeout: time.Second * 2, // Maximum of 2 secs
+	}
+
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return err
+	}
+
+	res, getErr := remoteClient.Do(req)
+	if getErr != nil {
+		return getErr
+	}
+
+	data, readErr := ioutil.ReadAll(res.Body)
+	if readErr != nil {
+		return readErr
+	}
+
+	return readAliasFromBytes(data, preTask)
+}
+
+/* Parses out local alias files and adds their content to the passed in
+PreTask. Note alias definitions already in preTask will not be
+overwritten. */
+func addAliasFromFile(preTask *PreTask, fileURI string) error {
+
+	data, fileReadingError := ioutil.ReadFile(fileURI)
+	if fileReadingError != nil {
+		return fileReadingError
+	}
+	return readAliasFromBytes(data, preTask)
+}
+
+/* Parses out alias  definitions from a given bytes array and appends
+them to the PreTask. Note alias definitions already in preTask will
+not be overwritten even if present in the array. */
+func readAliasFromBytes(data []byte, preTask *PreTask) error {
+
+	aliasMap := &map[string]string{}
+
+	if err := yaml.Unmarshal(data, aliasMap); err != nil {
+		return err
+	}
+
+	for key, value := range *aliasMap {
+		if _, ok := preTask.AliasMap[key]; !ok {
+			preTask.AliasMap[key] = value
+		}
+	}
+	return nil
+}
+
+// PreprocessString handles managing alias definitions from a provided string definitions expected to be in JSON format.
+func PreprocessString(preTask *PreTask, str string) (string, error) {
+	//preTask.loadGlobalDefinitions TODO?
+
+	// Load Remote/Local alias definitions
+	if externalDefinitionErr := preTask.loadExternalAlias(); externalDefinitionErr != nil {
+		return "", externalDefinitionErr
+	}
+	//Validate alias definitions
+	if improperFormatErr := preTask.resolveMapAndValidate(); improperFormatErr != nil {
+		return "", improperFormatErr
+	}
+	var out strings.Builder
+	var command strings.Builder
+	ongoingCmd := false
+
+	// Search and Replace
+	for _, char := range str {
+		if ongoingCmd {
+			//Maybe just checking if non alphanumeric, only allow alpha numeric aliases?
+			if strings.Contains(")}/ .,;]&|'~\n\t", string(char)) { // Delineates the end of an alias
+				resolvedCommand, commandPresent := preTask.AliasMap[command.String()]
+				if !commandPresent {
+					return "", errUnknownAlias
+				}
+
+				out.WriteString(resolvedCommand)
+				if char != preTask.directive {
+					ongoingCmd = false
+					out.WriteRune(char)
+				}
+				command.Reset()
+
+			} else {
+				command.WriteRune(char)
+			}
+		} else if char == preTask.directive {
+
+			if ongoingCmd { // Escape character triggered
+				out.WriteRune(preTask.directive)
+				ongoingCmd = false
+				continue
+			}
+
+			ongoingCmd = true
+			continue
+		} else {
+			out.WriteRune(char)
+		}
+	}
+
+	return out.String(), nil
+}
+
+// PreprocessBytes Handles files or byte encoded data that can be parsed through pre processing
```

Comment:
could you name all the functions and structs lowercase since they are private to this package and there's no reason to export them (yet)